homeassistant/components/media_player/pandora.py

```diff
@@ -0,0 +1,332 @@
+"""
+Component for controlling Pandora stations through the pianobar client.
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/media_player.pandora/
+"""
+
+import logging
+import re
+import os
+import signal
+from datetime import timedelta
+import shutil
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.components.media_player import (
+    SUPPORT_NEXT_TRACK, SUPPORT_PAUSE,
+    SUPPORT_TURN_OFF, SUPPORT_TURN_ON,
+    SUPPORT_SELECT_SOURCE, SERVICE_MEDIA_NEXT_TRACK, SERVICE_MEDIA_PLAY_PAUSE,
+    SERVICE_MEDIA_PLAY, SERVICE_VOLUME_UP, SERVICE_VOLUME_DOWN,
+    MediaPlayerDevice)
+from homeassistant.const import (STATE_OFF, STATE_PAUSED, STATE_PLAYING,
+                                 STATE_IDLE)
+from homeassistant import util
+
+REQUIREMENTS = ['pexpect==4.0.1']
+_LOGGER = logging.getLogger(__name__)
+
+# SUPPORT_VOLUME_SET is close to available but we need volume up/down
+# controls in the GUI.
+PANDORA_SUPPORT = \
+    SUPPORT_PAUSE | \
+    SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_NEXT_TRACK | \
+    SUPPORT_SELECT_SOURCE
+
+CMD_MAP = {SERVICE_MEDIA_NEXT_TRACK: 'n',
+           SERVICE_MEDIA_PLAY_PAUSE: 'p',
+           SERVICE_MEDIA_PLAY: 'p',
+           SERVICE_VOLUME_UP: ')',
+           SERVICE_VOLUME_DOWN: '('}
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=2)
+CURRENT_SONG_PATTERN = re.compile(r'"(.*?)"\s+by\s+"(.*?)"\son\s+"(.*?)"',
+                                  re.MULTILINE)
+STATION_PATTERN = re.compile(r'Station\s"(.+?)"', re.MULTILINE)
+
+
+# pylint: disable=unused-argument
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the media player pandora platform."""
+    pandora = PandoraMediaPlayer('Pandora')
+
+    # make sure we end the pandora subprocess on exit in case user doesn't
+    # power it down.
+    def _stop_pianobar(_event):
+        pandora.turn_off()
+
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _stop_pianobar)
+    add_devices([pandora])
+
+
+# pylint: disable=too-many-instance-attributes
+class PandoraMediaPlayer(MediaPlayerDevice):
+    """A media player that uses the Pianobar interface to Pandora."""
+
+    # pylint: disable=abstract-method
+    def __init__(self, name):
+        """Initialize the demo device."""
+        MediaPlayerDevice.__init__(self)
+        self._name = name
+        self._player_state = STATE_OFF
+        self._station = ''
+        self._media_title = ''
+        self._media_artist = ''
+        self._media_album = ''
+        self._stations = []
+        self._time_remaining = 0
+        self._media_duration = 0
+        self._pianobar = None
+        self._verify_pianobar()
+
+    def _verify_pianobar(self):
+        """Verify that Pianobar is properly installed."""
+        pianobar_exe = shutil.which('pianobar')
+        if not pianobar_exe:
+            raise RuntimeError('The Pandora component depends on the Pianobar '
```

Comment:
The exception is poorly timed as you still registered the stop listener. Also, don't throw an exception but instead just print a warning to `_LOGGER` and return `False` from setup_platform.